### PR TITLE
BUG: Jekyll/Jekyll Admin race condition remediation

### LINKED ISSUE #310: Liquid exception finding multiple files even though there's only one
I'm seeing what seems like it might be a race condition with the liquid exception handling for file includes:

https://gist.github.com/mtlynch/2de09ceeaf1c55193f33c81dcfe62e55

Every few saves, it gets into a state where it claims that multiple includes have the same title, but the titles seem to be configured correctly.

The errors are also accompanied by lines like this:

```
/usr/local/lib/ruby/gems/2.4.0/gems/jekyll-3.8.2/lib/jekyll/reader.rb:42: warning: conflicting chdir during another chdir block
```

which makes me think there's maybe a race condition (although I've been seeing Jekyll produce that warning, seemingly at random, for many months now).

This is on Ubuntu 16.

```
$ ruby --version
ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-linux]
```

### PR DESCRIPTION
**Pull request by shredtechular**
Fixes #310 

This PR includes a workaround for the problem pointed out in #310.  Unfortunately, attempts at finding the root cause have not been fruitful, however, since this issue is:
- only reproducable under certain conditions AND while using `Jekyll-admin` to alter content 
- and is not a problem when building the production site
I feel like this workaround should be sufficient to eliminate the bigger problems of this bug.

## What this workaround WILL do
This workaround will:
- [X] Ensure there are no false positives for duplicate file include titles by creating an array of file paths and checking to ensure the same file path is not "counted" as a duplicate file title
- [X] Provide the intended and original functionality for pro-actively finding duplicate and/or incorrect file include titles

## What this workaround DOESN'T do
This workaround doesn't:
- [ ] Address the `conflicting chdir during another chdir block` build warning.  This appears to be at the root of the issue, but again, only occurs when using `Jekyll-admin` for editing content.  So, I believe, the race condition is still present, but doesn't affect how the site is built either for production or when using `Jekll-admin`.

## Attempts at finding the root of the race condition
I have tried several different things to try and get to the root of the race condition unsuccessfully.  Here are things I have tried:
- Created a very basic Jekyll site with a "files" collection.  Could not reproduce the race condition behavior however thus far on the most basic of Jekyll sites.  It should be noted that build times are significantly faster with such a basic site however.  Based on further attempts, I believe the condition mostly presents itself when build times are over a couple of seconds, otherwise, very difficult to reproduce.
- After the basic site, I then started trying to add mtlynch.io functionality piece-by-piece, but couldn't reproduce until adding more and more posts or more and more functionality.  So, I don't think there is anything in our site's code that is causing this, I believe, this has more to do with build times, but it's very difficult to get any consistency with this.
- Built the site using the initial commit of mtlynch.io to see if there was something we introduced along that way that caused this, however, even with that build, I saw the `conflicting chdir during another chdir block` build warnings using `Jekyll-admin` so adding collections and functionality does not seem to have caused this.
- Tried updating Ruby versions, Jekyll, Jekyll-responsive-image, and other gems, but no effect on the condition.  Could be some other benefits that I will try to get a PR together for though.

## Additional Notes
- [X] **There was a broken link in the `2017-05-20-windows-sia-mining.md` post, so I added an ignore to the HTMLProofer parameters to get the build successful**
- [X] **Please note** this PR does change the "warning" liquid exceptions back to "error" exceptions

I can continue to try and identify what the root cause issue is with this, but this could be very time consuming and wanted to get the workaround in for now at least so intended functionality isn't limited.  My gut feeling is this has something to do with the Jekyll "watch" process interferring with the spawned Jekyll-admin sinatra servers also "watching" files/directories and both processes attempting to access the same files at the same time.  From what I've read it seems the `chdir` command isn't thread safe in Ruby so that is causing the problem.

---
This change is [Reviewable](https://reviewable.io/reviews/mtlynch/mtlynch.io/316)

### CODE REVIEW

File: _includes/files.html [hunk 1]
```diff
@@ -16,23 +16,25 @@
...
+  {% assign lang_highlight = include.language %}
 {% endif %}
 
 {% comment %}
-<!-- initialize file count -->
+<!-- initialize file count & array -->
 {% endcomment %}
 {% assign file_count = 0 %}
+{% assign found_file_array = " " | split: ',' %}
+{% assign found_file_array =  found_file_array | shift %}
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Yeah, I can see where that looks to be not doing a lot... what I want there is an empty array, but you can't create an empty array in liquid without using the `split` filter (that I know of and research turned up at least...).  So, what that is doing is creating an array of 1 item which is an empty space.  Then using `shift` to remove that item, thus giving me something more like:

{% assign found_file_array = [] %}

I can add some comments there to explain more...

[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Added comment in a57d751dd13ba53a48ee253b64866009e6989db8

File: _includes/files.html [hunk 2]
```diff
@@ -52,6 +54,8 @@
   {% assign subdir = subdir_array[1] %}
 
   {% if subdir == page.slug %}
+    {% if found_file_array.size == 0 %}
+      {% assign found_file_array =  found_file_array | push: file_path %}
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Good catch, fixed both lines and found a 3rd as well.  They have been fixed now in 88819f69b7f1d30b348a910fec07967709948e7c

File: _includes/files.html [hunk 3]
```diff
@@ -61,27 +65,40 @@
...
-    | append: include.title 
-    | append: ".  Please check the file include's title and directory again." 
-    | raise_exception: "warning" }}
-{% endif %}
+{% assign file_count = found_file_array | size %}
+
+{% unless file_count == 1 %}
+  {% if file_count == 0 %}
+    {% comment %}
+    <!-- if there has been NO files found in the subdirectory that match the
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Fixed in 88819f69b7f1d30b348a910fec07967709948e7c

File: _includes/files.html [hunk 4]
```diff
@@ -61,27 +65,40 @@
...
+         title, then raise an error -->
+    {% endcomment %}
+    {{ "Zero file includes matched the title, "
+      | append: include.title
+      | append: ".  Please check the file include's title and directory again."
+      | raise_exception: "error" }}
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: OK, changed them back to "warning" instead of "error" in 88819f69b7f1d30b348a910fec07967709948e7c

File: _tests/build [hunk 5]
```diff
@@ -32,7 +32,7 @@ bundle exec htmlproofer ./_site \
   --allow-hash-href \
   --empty-alt-ignore \
   --url-swap "http\://localhost\:4000/:/" \
-  --url-ignore "/vimeo.com/,/upwork.com/" \
+  --url-ignore "/vimeo.com/,/upwork.com/,http://amzn.to/2rBvkbA" \
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _tests/build]: Ah okay, no worries.  That has been removed now in 88819f69b7f1d30b348a910fec07967709948e7c.

**Comment by shredtechular** [comment]
OK, added some additional comments to `_includes/files.html` that hopefully explain some of the logic in better detail and also tried to point out where the "workaround" was required.  It's probably a poor choice of words on my part to use "workaround" as that makes it feel hacky to me.  To be clear the additional logic that is included in this file is **required** due to the Jekyll admin bug since it worked fine without using Jekyll admin.  However, I almost feel like this logic may be a little more full proof and adds an additional layer of checking so if you accidentally copied the same include file into the same directory, it wouldn't mess the page up.  I assume that would never happen!  But...maybe this is a more encompassing/better solution regardless if it was created only to deal with a bug in Jekyll admin?  Just wanted to add that in after reading "workaround" out loud...maybe I'm thinking too much into it! haha

File: _includes/files.html [hunk 6]
```diff
@@ -30,11 +30,18 @@
...
-<!-- initialize file count & array -->
+<!-- initialize file count -->
 {% endcomment %}
 {% assign file_count = 0 %}
+
+{% comment %}
+<!-- create an empty array to store UNIQUE file paths that match the correct
+  file path and title.  Unfortunately due to Jekyll Admin bug, this array of 
+  UNIQUE files is part of the workaround to ensure no false positives are
+  thrown for duplicate title error catching. -->
```
Comment (mtlynch) [inline _includes/files.html]: Can we lowercase "unique"? The uppercase makes it look a little bit like an acronym.

I have a hard time understanding the second sentence. It looks like there's maybe a word missing. It also uses passive voice ("are thrown"), which makes it ambiguous because it omits the subject. Suggest rephrasing to remove active voice.
(ditto throughout)

The phrase "duplicate title error catching" is a bit hard to parse. Can we say "catching duplicate filenames"?
(ditto throughout)

From the high-level comment, it sounds like this isn't specifically a workaround, just a more robust implementation that works regardless of whether there's a bug in Jekyll Admin. If that's the case, we should just omit the "workaround" from the discussion.
(ditto throughout)

File: _includes/files.html [hunk 7]
```diff
@@ -68,9 +82,17 @@
...
+        file title error check will trigger below.  This is only needed as part
+        of a workaround to ensure the same file is counted twice for duplicate
+        title error trapping. -->
+      {% endcomment %}
       {% for found_file in found_file_array %}
         {% if found_file != file_path %}
-          {% assign found_file_array =  found_file_array | push: file_path %}
+          {% assign found_file_array = found_file_array | push: file_path %}
         {% endif %}
       {% endfor %}
```
Comment (mtlynch) [inline _includes/files.html]: I'm having trouble understanding this logic.

If `found_file_array` is `['A', 'B', 'C']` and `file_path` is `'C'`, won't it do an iteration for `'A'` and push `'C'`, so `found_file_array` is `['A', 'B', 'C', 'C']`? And then do the same thing when it iterates for `'B'`?

I'd rather not modify an array while we're iterating it because it's hard to reason about what will happen. Unless the developer is very familiar with the implementation details of liquid loop iteration, it's unclear if the loop iterates over elements added to the array *during* the loop.

File: _includes/files.html [hunk 8]
```diff
@@ -30,11 +30,18 @@
...
+
+{% comment %}
+<!-- create an empty array to store UNIQUE file paths that match the correct
+  file path and title.  Unfortunately due to Jekyll Admin bug, this array of 
+  UNIQUE files is part of the workaround to ensure no false positives are
+  thrown for duplicate title error catching. -->
+{% endcomment %}
 {% assign found_file_array = " " | split: ',' %}
-{% assign found_file_array =  found_file_array | shift %}
+{% assign found_file_array = found_file_array | shift %}
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Based on this comment and the other one about setting it to a single spaced string, I decided to re-work this logic to simply initialize a string and then split it into an array only when necessary as opposed to starting off with an array.  Creating a 0 length array is just a little awkward in liquid so hopefully this new implementation is more readable.

The new logic is in 0df3a12 and makes this comment N/A now.

File: _includes/files.html [hunk 9]
```diff
@@ -30,11 +30,18 @@
...
 {% endcomment %}
 {% assign file_count = 0 %}
+
+{% comment %}
+<!-- create an empty array to store UNIQUE file paths that match the correct
+  file path and title.  Unfortunately due to Jekyll Admin bug, this array of 
+  UNIQUE files is part of the workaround to ensure no false positives are
+  thrown for duplicate title error catching. -->
+{% endcomment %}
 {% assign found_file_array = " " | split: ',' %}
```
[the remark this answers is not in the document]
Comment (shredtechular) [reply _includes/files.html]: Please see above comment, this is now N/A in 0df3a12

File: _includes/files.html (hunk 6 above)
Comment (shredtechular) [reply _includes/files.html]: Updated this comment specifically to simply:

```"initialize an empty string to store unique file paths that match the correct file path and title."```

The logic changed, so previous comment is incorrect anyway, but I think this is more succinct and uses the active voice instead of passive voice?  Also reviewed/updated other comments to ensure active voice...I think... I suck at English!! ha!

Also agreed on removing the "workaround" from the comments... I don't think it is really necessary either given the point you made.

Fixed in 5a3cf23.

File: _includes/files.html (hunk 7 above)
Comment (shredtechular) [reply _includes/files.html]: Very valid points here...  Once we had an array with a length > 1, it really doesn't matter because it will throw the error as expected and needed.  However, I can see how modifying an array while iterating over it is not a good practice at all.

So, I've re-worked this logic so that it splits a string into the array and just sets a variable if it is NOT found.  Then appends that "not found" value to the string after the iteration has been complete.

So, I think this is better?  Its definitely more wordy/lines, but should be safer...  Since we are comparing file paths, it would be awesome to just see if the string "contains" the active file, but I really just worry that "contains" wouldn't be safe in ALL situations.  So, I think iterating over an array is the most comprehensive solution.

Fixed in 5a3cf23

**Comment by shredtechular** [comment]
Based on comments, adjusted this logic again overall.  This should be ready for another round of review when ready.
